### DISCUSSION TITLE
How to disable strict https?

### DISCUSSION THREAD

**Discussion by gardellajuanpablo** [keycloak/keycloak#10740] [Archive]
Hi,

I am trying to be able to use Keycloak X with http only. First access works fine, but when I am trying to access to admin, it always append https. Tried different combinations and by looking the logs and the code, there is a configuration that appends https at this [line](https://github.com/keycloak/keycloak/blob/0a9387ff4f121600fbb6acb1f6f89c3fc9fce76d/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/hostname/DefaultHostnameProvider.java#L205). How I can disable it? I put `quarkus.conf` with `strict-https=false` but does not work, Any idea?

``
         | 2022-03-14 12:46:29,094 INFO  [org.keycloak.quarkus.runtime.hostname.DefaultHostnameProvider] (main) Hostname settings: FrontEnd: <request>, Strict HTTPS: true, Path: /auth, Strict BackChannel: false, Admin: <request>
``

**Comment by jabbasi** [comment]
You can follow some tips here: https://github.com/keycloak/keycloak/issues/10627.

**Comment by dasniko** [comment]
Undocumented config parameter: `hostname-strict-https=false`
Or, as environment variable: `KC_HOSTNAME_STRICT_HTTPS=false`

**Reply by daurrutia** [reply]
Thank you for this and for all your contributions to the community.

**Reply by omasseau** [reply]
Is it deliberate that this config parameter is undocumented or is it an oversight in the documentation ?

**Reply by codespearhead** [reply]
This is a deliberate decision as concluded in #12875 , because what people asking this question problably want is not to disable `HTTPS`, which is a huge security risk, but to run Keycloak behind a reverse proxy which will deal with SSL/TLS instead of Keycloak itself, as explained [here](https://github.com/keycloak/keycloak/discussions/28279#discussioncomment-9001818).

@gardellajuanpablo can you consider closing this discussion?

**Reply by raffaeler** [reply]
@codespearhead Another common scenario is where you have to implement oAuth features by yourself.
If you need to debug what's happening on the transport, it is definitely desirable to disable `https` to log and analyze the traffic.

While I totally agree that `HTTPS` must be always used even in LAN-only scenarios, there are a number of reasonable reasons to disable it. 
Also, IMO, reverse-proxy is not enough to justify disabling `HTTPS` even if it is certainly a popular reason.

**Reply by gardellajuanpablo** [reply]
Yes, thank you.

**Reply by gardellajuanpablo** [reply]
I was not able to disable it, I had to keep it enable.

**Reply by vmuzikar** [reply]
In Keycloak 25 we're introducing a new Hostname experience that (among other changes) removes the hostname-strict-https option. You can give it a try already in our [nightly builds](https://github.com/keycloak/keycloak/releases/tag/nightly). See [upgrading guide](https://github.com/keycloak/keycloak/blob/main/docs/documentation/upgrading/topics/changes/changes-25_0_0.adoc#new-hostname-options) and [docs](https://github.com/keycloak/keycloak/blob/9e10cec665ab48a7b4f53f8f70560a5a3a2d153d/docs/guides/server/hostname.adoc).

**Reply by raffaeler** [reply]
> In Keycloak 25 we're introducing a new Hostname experience that (among other changes) removes the hostname-strict-https option. You can give it a try already in our [nightly builds](https://github.com/keycloak/keycloak/releases/tag/nightly). See [upgrading guide](https://github.com/keycloak/keycloak/blob/main/docs/documentation/upgrading/topics/changes/changes-25_0_0.adoc#new-hostname-options) and [docs](https://github.com/keycloak/keycloak/blob/9e10cec665ab48a7b4f53f8f70560a5a3a2d153d/docs/guides/server/hostname.adoc).

No "dev mode" anymore? Do you support logging the traffic using `SSLKEYLOGFILE` at least?
If this feature also impacts HSTS, please consider avoiding enforcing HSTS on localhost as this causes a huge amount of troubles.

**Reply by vmuzikar** [reply]
> No "dev mode" anymore?

`start-dev` is still there with [relaxed validations](https://github.com/keycloak/keycloak/blob/9e10cec665ab48a7b4f53f8f70560a5a3a2d153d/docs/guides/server/hostname.adoc?plain=1#L134) for hostname.

> Do you support logging the traffic using `SSLKEYLOGFILE` at least?

Can you please elaborate how do you expect it to work with Keycloak? If I understand correctly, this is something that works on browser level.

> If this feature also impacts HSTS, please consider avoiding enforcing HSTS on localhost as this causes a huge amount of troubles.

HSTS is still set on Realm level, it has nothing to do with the Hostname provider. Please see my [previous comment](https://github.com/keycloak/keycloak/issues/14162#issuecomment-1784901028).

**Reply by raffaeler** [reply]
> Can you please elaborate how do you expect it to work with Keycloak? If I understand correctly, this is something that works on browser level.
> 
> > If this feature also impacts HSTS, please consider avoiding enforcing HSTS on localhost as this causes a huge amount of troubles.
> 
> HSTS is still set on Realm level, it has nothing to do with the Hostname provider. Please see my [previous comment](https://github.com/keycloak/keycloak/issues/14162#issuecomment-1784901028).

HSTS is enforced on the browser **on request** by the server.
It is the server that asks the browser to refuse any unsecure connection on a domain.

This means that if you deploy KC with HSTS on localhost, the browser will associate HSTS on **any** localhost connection, even if related to a different server/app/appliance.

Since localhost is clearly a dev/test scenario only, I am trying to avoid this mess when demoing or developing apps locally with KC.

Same for http/https. When serving requests on localhost I eould love to have an option to avoid secure transport.

**Reply by jonkoops** [reply]
You can perhaps use a `.local` domain? These are considered secure by browsers for development, and should have a unique domain. E.g. `foo.local` & `bar.local`.

**Reply by raffaeler** [reply]
This is what I am actually doing with an external router, but certain times you have to use localhost because some libraries provide specific features only on localhost.

**Reply by codespearhead** [reply]
Can you list a few of those libraries?

**Reply by raffaeler** [reply]
The first coming into my mind is mongodb that has exception for `localhost`. But I also have to work with a custom IP from a customer that also have special rules for `localhost`.
Also, while the `.local` domain can open a vulnerability on a LAN, this can't happen on `localhost`.

**Comment by gardellajuanpablo** [comment]
If I enable this option, accessing with https fails (when it is behind proxy). I guess that maybe is expected, it will not work in both modes.

`2022-03-29 00:36:07.526 ERROR 1 --- [nio-8080-exec-7] o.k.adapters.OAuthRequestAuthenticator   : failed verification of token: Invalid token issuer. Expected 'http://foo.com/auth/realms/realmId', but was 'https://foo.com/auth/realms/realmId'`

Anyway, the option you mentioned worked but I have to stick with default configuration.

**Reply by andreaTP** [reply]
Have you tried to correctly configure the proxy option?
If I read those requirements correctly you might want to setup: `KC_PROXY=passthrough`

**Reply by gardellajuanpablo** [reply]
I will check and let you know, thanks for the suggestion.

**Comment by vmuzikar** [comment]
Please note that we're currently also facing some issues with regards to HTTPS, hostnames and proxy settings. See #11579 and #11581.

**Reply by vmuzikar** [reply]
CC: @DGuhr

**Comment by raffaeler** [comment]
I disabled HSTS from the dockerfile with:
```
ENV KC_HOSTNAME_STRICT_HTTPS=false
```

When the container starts, I see this in the log:

```
2022-09-01 19:36:44,991 INFO  [org.keycloak.quarkus.runtime.hostname.DefaultHostnameProvider] (main) Hostname settings: 
   FrontEnd: <request>,
   Strict HTTPS: false, Path: <request>,
   Strict BackChannel: false,
   Admin: <request>,
   Port: -1,
   Proxied: false
```

Anyway, after visiting the admin console, the chrome page `chrome://net-internals/#hsts` tells me that HSTS is enabled for localhost (yes, I published KC on localhost in container to test a bunch of scenarios).

Is there any other effective ways to **disable HSTS**? HSTS on localhost causes a huge amount of problems.

TIA

**Reply by vmuzikar** [reply]
This sounds like a possible bug. Can I ask you to create a GH Issue for it? Thanks.

**Reply by raffaeler** [reply]
Thanks @vmuzikar. I filed a new issue [here](https://github.com/keycloak/keycloak/issues/14162).

**Reply by codespearhead** [reply]
Should this discussion be closed now that Keycloak v25.0.0 is out?

https://github.com/keycloak/keycloak/blob/a8aa410ad350f550bc60e82f4beda32eb064528f/docs/documentation/upgrading/topics/changes/changes-25_0_0.adoc?plain=1#L32-L34
